htlcswitch/payload/hop.go

```diff
@@ -0,0 +1,84 @@
+package payload
+
+import (
+	"io"
+
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/record"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+// Hop encapsulates all information delivered in an onion payload. A Hop can
+// represent either a TLV or legacy payload. The primary forwarding instructsion
+// can be accesed via ForwardingInfo, and additional records can be accessed by
+// other member functions.
+type Hop struct {
+	// FwdInfo holds the basic parameters required for HTLC forwarding, e.g.
+	// amount, cltv, and next hop.
+	FwdInfo ForwardingInfo
+
+	// MPP holds the info provided in an option_mpp record when parsed from
+	// a TLV onion payload.
+	MPP *record.MPP
+}
+
+// NewLegacyHop builds a Hop from the amount, cltv, and next hop parameters
+// provided by leegacy onion payloads.
+func NewLegacyHop(amt uint64, cltv uint32, nextHop lnwire.ShortChannelID) *Hop {
+	return &Hop{
+		FwdInfo: ForwardingInfo{
+			Network:         BitcoinNetwork,
+			NextHop:         nextHop,
+			AmountToForward: lnwire.MilliSatoshi(amt),
+			OutgoingCTLV:    cltv,
+		},
+	}
```

Comment:
Shouldn't deserialization of the legacy hop bytes happen in here as well? (tlv is also deserialized in this object)

Reply:
done

Reply:
I actually meant `HopData.Decode` that is now in `lightning-onion`

Reply:
ah i see, maybe we can add a todo since it is a bit more involved